method/stLearn/method_stLearn.py

```diff
@@ -0,0 +1,235 @@
+            )
+
+seed = args.seed
+## Your code goes here
+import random
+import numpy as np
+import pandas as pd
+import stlearn as st
+
+def get_anndata(args):
+    import anndata as ad
+    import scipy as sp
+    X = sp.io.mmread(args.matrix)
+    if sp.sparse.issparse(X):
+        X = X.tocsr()
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    # Filter
+    if "selected" in observations.columns:
+        X = X[observations["selected"].to_numpy().nonzero()[0], :]
+        observations = observations.loc[lambda df: df["selected"]]
+    if "selected" in features.columns:
+        X = X[:, features["selected"].to_numpy().nonzero()[0]]
+        features = features.loc[lambda df: df["selected"]]
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(
+        X=X, obs=observations, var=features, obsm={"spatial_pixel": coordinates}
+    )
+
+    adata.obs["imagecol"] = adata.obsm["spatial_pixel"][:, 0]
+    adata.obs["imagerow"] = adata.obsm["spatial_pixel"][:, 1]
+    adata.obs["array_row"] = adata.obs["row"]
+    adata.obs["array_col"] = adata.obs["col"]
+    # Use morphology information
+    # Modified from stLearn.adds.add_image.py
+    if args.image is not None:
+        image_tif = args.image[0]
+        image_json = args.image[1] if len(args.image) == 2 else None
+        from PIL import Image
+        # Add image
+        Image.MAX_IMAGE_PIXELS=None
+        im = Image.open(image_tif)
+        image = np.array(im)
+        spatial_key = "spatial"
+        library_id = "sample"
+        quality = "hires"
+        try:
+            adata.uns[spatial_key] = {library_id: {}}
+            adata.uns[spatial_key][library_id]["images"] = {}
+            adata.uns[spatial_key][library_id]["images"][quality] = image
+            adata.uns[spatial_key][library_id]["use_quality"] = quality
+            # if technology == 'Visium':
+            #     adata.uns[spatial_key] = {library_id: {}}
+            #     adata.uns[spatial_key][library_id]["images"] = {}
+            #     adata.uns[spatial_key][library_id]["images"][quality] = image
+            #     adata.uns[spatial_key][library_id]["use_quality"] = quality
+            if technology != 'Visium':
+                adata.uns[spatial_key][library_id]["scalefactors"] = {}
+                if image_json is not None:
+                    with open(image_json) as json_file:
+                        scale_factors = json.load(json_file)
+                    scale = scale_factors.get('scale', 1.0)
+                    spot_diameter_fullres = scale_factors.get('spot_dimeter_fullres', 50.0)
+                else:
+                    scale = 1.0
+                    spot_diameter_fullres = 50.0
```

Comment:
I am wondering whether we should rather fail than trying to guess parameters?

Reply:
Yes, but in the H_E.json file for LIBD_DLPFC data, it only provides the scale parameter...
Should we handle failures and return a message requesting such parameters?